tests/drivers/clock_control/pwm_clock/src/main.c

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2023 Andriy Gelman <andriy.gelman@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/ztest.h>
+
+#define NODELABEL DT_NODELABEL(samplenode)
+
+ZTEST(pwm_clock, test_pwm_clock_rate_change)
+{
+	const struct device *clk_dev = DEVICE_DT_GET(DT_CLOCKS_CTLR(NODELABEL));
+	int ret;
+	uint32_t target_rate, clock_rate;
+
+	printk("PWM-based clock control\n");
+	zassert_equal(device_is_ready(clk_dev), true, "%s: PWM clock device is not ready",
+		      clk_dev->name);
+
+	ret = clock_control_on(clk_dev, 0);
+	zassert_equal(0, ret, "%s: Unexpected err (%d) from clock_control_on",
+		      clk_dev->name, ret);
+
+	printk("Started clock\n");
+
+	ret = clock_control_get_rate(clk_dev, 0, &clock_rate);
+	zassert_equal(0, ret, "%s: Unexpected err (%d) from clock_control_get_rate",
+		      clk_dev->name, ret);
+
+	printk("Clock frequency set to %dHz\n", clock_rate);
+
+	k_sleep(K_SECONDS(10));
```

Comment:
Please do not sleep for 10 seconds in a Ztest unless absolutely required. This will prolong test runs for everyone.

Reply:
I removed the sleep (I had it before so that I could check with an oscilloscope that the frequency was changing).